채팅룸 검수

ChoMk opened the pull request:
여기서 코멘트 달기 편해서 여기다가 달겡

bsgreentea committed:
Implement #9 데이터베이스 구성
- dao 구현
- typeconverter를 사용하려 했으나 Message에 room id 부여

bsgreentea committed:
Implement #9 메세지 레포지토리 구성

bsgreentea committed:
Implement #9 메세지 정렬 기준 수정

bsgreentea committed:
Implement #9 메세지 어댑터 구성
- 채팅방 리사이클러뷰 구성
- 채팅방 레이아웃과 연결

bsgreentea committed:
Style #9 채팅방 메세지 아이템 데이터바인딩

bsgreentea committed:
Refactor #9 메세지 primary key 설정

bsgreentea committed:
Implement #9 채팅방 구성
- 보내기 버튼 데이터바인딩
- usecase 어떻게 하나요ㅜㅜ

ChoMk commented on app/src/main/java/com/example/messenger/database/message/MessageDao.kt:
```diff
@@ -0,0 +1,22 @@
+    fun insertMessage(message: Message): Completable
+
+    @Query("SELECT * FROM messages WHERE roomID = :roomID ORDER BY messageTime")
+    fun getMessageList(roomID: Int): Single<List<Message>>
```
메시지방 입장 시간을 저장(따로 데이터를 만들어야하는지에 대해서는 고민해봐야 할 듯)하고 이를 이용해서 그 이후의 특정 메시지 수(아직 정하지는 않았지만 예를 들면 한번 로딩에 50개 정도의 메시지를 불러오게..)를 불러오면 좋을 듯 ㅎㅎ

ChoMk commented on app/src/main/java/com/example/messenger/repository/message/MessageRepository.kt:
```diff
@@ -0,0 +1,13 @@
+ */
+interface MessageRepository {
+    fun insertMessageToLocal(msg: Message): Completable
+    fun getMessageListFromLocal(roomID: Int): Single<List<Message>>
```
만약 시간이 들어가면은 여기에 추가적인 프로퍼티가...

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+
+    init {
+        subscribeEvent()
+        messageToSend = ObservableField("")
```
어차피 초기화면은 위쪽으로 옮겨서private var messageToSend: ObservableField<String> = ObservableField("") 이렇게 해줘도 init에서 초기화하는건 같아

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    fun getText(): ObservableField<String>{
+        return messageToSend
```
이 부분을 지우고 messageToSend를 직접 연결할 수 있게 변수명을 조정하면 좋지 않을까? messageContent... 그리고 getText가 아니라 바로 퍼블릭으로 바로 바인딩 시키던지 아니면 커스텀 게터와 세터를 저 messageToSend에 만들어 주면은 좋을 듯

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/adapter/MessageReceivedViewHolder.kt:
```diff
@@ -0,0 +1,17 @@
+ */
+
+class MessageReceivedViewHolder(
+    val binding: ItemMessageReceivedBinding
```
아마 ide 잘 설절했으면 노랭이 줄이 뜰꺼얌.. 그런 노랭이 줄도 지워주려고 노력하는 걸 추천

bsgreentea replied:
노랭이 안뜨던데... 으윽.... 혹시 private 때문인가요?

ChoMk replied:
ㅇㅇㅇ

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/adapter/MessageSentViewHolder.kt:
```diff
@@ -0,0 +1,16 @@
+class MessageSentViewHolder(
+    val binding: ItemMessageSentBinding
+) : RecyclerView.ViewHolder(binding.root) {
+    fun bind(message: Message) {
```
지금보면 message viewholder가 여러가지가 있자너 이 부분에서 공통된 행동들을 base message viewholder로 interface로 묶는거는 어떨까?... 이게 지금 당장은 bind만 작성했지만... 여를 들어서 view holder가 엄청 여러개가 생길 수 있자너 나중에는...
이미지 홀더 *2 , 파일 홀더 *2, 메시지 *2, 등등... 만약에 그런 뷰 홀더들이 베이스로 묶이지 않은 상태에서 뷰 홀더에 공통된 기능을 넣고 싶다고 할 경우에 저 많은 뷰 홀더들에 실수로 하나를 공통된 기능을 작성하지 않고 넘어갈 수 있지... (거의 똑같은 경우라면은 구현체를 인터페이스에 넣을 수도 있고...)(예를 들자면 롱 클릭시에 삭제 기능? 그러면 인터페이스에 fun onLongClick을 넣어주면은... 머 이런식?)

ChoMk commented on app/src/main/res/layout/activity_chatting_room.xml:
```diff
@@ -4,54 +4,68 @@
+            android:hint="@string/chat_hint"
+            android:maxLines="6"
+            android:minHeight="48dp"
+            android:text="@={viewModel.text}"
```
바로 viewmodel 내부의 변수로 바인딩하는게 어떨까 싶어 다른사람이 읽기에도 그게 편할거 같고... 뭔가 후처리가 필요한 경우라면 커스텀 게터 세터를 쓰고,.. (위에 코멘트랑 동일)

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+        messageToSend = ObservableField("")
+    }
+
+    fun onClickSend() {
```
이 함수명을 sendMessageToServer이런식으로 하는건 어떨까? 뒤에 xml을 보니깐 onClick이 수행된 이후에 발생하는 함수 같은데... android:onClick="@{()->viewModel.onClickSend()}" -> android:onClick="@{()->viewModel.sendMessageToServer()}"
이런식으로 바꾼다면... 처음에는 저 위에 로직만 보고는 클릭시에 센드하기는 하는데 센드를 어디로하는지는 잘 모르자너... 그래서 이런식으로 바꾸는건 어떨까 싶어... 그리고 server로 send하는 이벤트가 onClick 시에만(그럴 수 있지만...) 발생하지는 않을 수 있자너?... 그런경우를 고려하여 함수명을 바꾸는건 어떨까 싶어

bsgreentea replied:
이렇게 하는게 훨씬 명확하네요 굿굿

ChoMk commented on app/src/main/res/layout/item_message_received.xml:
```diff
@@ -3,58 +3,67 @@
+            android:maxWidth="240dp"
+            android:padding="8dp"
+            android:text="@{receivedItemViewModel.messageContent}"
+            android:textColor="#ffffff"
```
res value에 정의하면은 좋을거 같아

ChoMk committed:
Implement #9 stomp 로직 구상
- 라이브러리 추가
- message repo에 필요한 로직 추가
- convert 로직 필요...(선언만)

ChoMk committed:
Implement #9 변환함수 구현
- stomp 전송시에 string으로 전달하기 때문에 convert 구성

ChoMk committed:
Fix #9 빌드오류 수정
- enum class를 위한 typeconvertor추가...

bsgreentea committed:
Implement #9 채팅방 usecase

bsgreentea committed:
Implement #9 채팅방 event

bsgreentea committed:
Implement #9 usecase 수정
- 최근 50개 메시지를 가져오도록 수정

bsgreentea committed:
Refactor #9 피드백 반영

ChoMk committed:
#9 Fix Build 오류 수정
Merge commit '92bb7e6d777e268ac7da6ff5f05ced3ac2adbab1' into feature/#9_chatting_room_socket_impl

# Conflicts:
#	app/src/main/java/com/example/messenger/repository/message/MessageRepository.kt
#	app/src/main/java/com/example/messenger/repository/message/MessageRepositoryImpl.kt

ChoMk committed:
Implement #9 유즈케이스 생성
- SendMessageUseCase
- ReceiveMessageUseCase

bsgreentea committed:
Refactor #9 usecase 수정

bsgreentea committed:
Refactor #9 id 수정
- roomID -> roomId

bsgreentea committed:
Refactor #9 friends usecase 반영
- ChoMK 수정사항 반영

bsgreentea committed:
Refactor #9 채팅방 메세지 로딩 쿼리문 수정
- 최근 메세지를 50건씩 불러오도록 수정

ChoMk committed:
Test #9 소켓 테스트
- 테스트

bsgreentea committed:
Implement #9 어댑터 수정
- 채팅방 수정

bsgreentea commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomActivity.kt:
```diff
@@ -23,5 +27,17 @@ class ChattingRoomActivity : BaseActivity() {
+//        chattingRoomViewModel.loadMessageUseCase.loadMessages()
+    }
+
+    private fun subscribeMessageList(messageAdapter: MessageAdapter) {
+        chattingRoomViewModel.messageList.observe(this, Observer {
+            messageAdapter.submitList(it.toMutableList())
+        })
```
@ChoMk  이 부분 어떻게 해야 맞는건지 잘 모르겠네요... 설명 부탁드립니다ㅜㅜ

ChoMk replied:
로직을 설명하면은 뷰모델에 메시지 데이터를 엑티비티에서 구독하고 변경이 일어나면 엑티비티가 뷰모델에 있는 메시지 리스트를 구독하고 있으니 그에대한 변경값을 받고 어뎁터로 갱신(일반적인 어뎁터는 아이템에 넣고 아이템 체인지 노티파이를 날리는데 여기서는 그냥 서브밋 리스트로 넘겨주면 그에 맞는 변경점을 찾아서 갱신)시켜주는 로직임

ChoMk replied:
뷰모델에서는 뷰를 갖고있을 수 없으니 이런 방식으로 진행했는데... 흠

ChoMk replied:
아님 어뎁터를 뷰모델에 주입하는 방식도 있는데...

bsgreentea committed:
Fix #9 roomId 변수형 변환
- Int -> String

bsgreentea committed:
Implement #9 리스너 생성
- 메세지 리스트를 채팅방에서 최상단으로 올릴 경우 메세지 더 로드하기
- 함수를 어떻게 써야할지 모르겠음....

ChoMk commented on app/src/main/java/com/example/messenger/usecase/LoadMessagesUseCase.kt:
```diff
@@ -0,0 +1,43 @@
+                }
+                .subscribe()
+        )
+    }
```
private fun loadLatestFiftyMessages(roomId: String, from: Int) {
        disposables.add(
            messageRepository.getLatestFiftyMessages(roomId, from)
                .doOnSuccess {
                    it.forEach { message ->
                        ChattingRoomEvent.addMessageToList(message)
                    }
                }
                .subscribe()
        )
    }

ChoMk replied:
```suggestion
    }
````` private fun loadLatestFiftyMessages(roomId: String, from: Int) {
        disposables.add(
            messageRepository.getLatestFiftyMessages(roomId, from)
                .doOnSuccess {
                    it.forEach { message ->
                        ChattingRoomEvent.addMessageToList(message)
                    }
                }
                .subscribe()
        )
    }`suggestion
    }
```

ChoMk replied:
이렇게 수정 가능

ChoMk replied:
foreach 사용

ChoMk commented on app/src/main/res/layout/activity_chatting_room.xml:
```diff
@@ -40,7 +40,7 @@
             android:maxLines="6"
             android:minHeight="48dp"
-            android:text="@={viewModel.text}"
+            android:text="@{viewModel.text}"
```
=이거 빠져도 됨? 에딧 텍스트 갖고오는게 =들어 갔던거 같아서 ㅎㅎㅎ

ChoMk committed:
Fix #9 통신오류 수정

bsgreentea committed:
Fix #9 코드 수정

ChoMk committed:
Refactor #9 소켓을 사용하는 엑티비티 분리
- connect와 disconnect를 엑티비티 resume, pause에서 하도록 수정

bsgreentea committed:
Implement #9 메세지 더 불러오기 구성
- 채팅방 내에서의 메세지 리스트에 대한 스크롤 리스너 구성
- 채팅방에서 스크롤을 아래로 당기면 더 메세지를 더 불러오도록 구성

bsgreentea committed:
Fix #9 잘못한 부분 되돌려놓음

bsgreentea commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomViewModel.kt:
```diff
@@ -16,11 +16,13 @@ class ChattingRoomViewModel(
     }
 
+//    private val chatListViewModel = ChatListViewModel(loadMessageUseCase)
+
```
@ChoMk 
메세지 리스트에 사용할 스크롤 리스너에 loadMessageUseCase와 roomId가 필요해보입니다.
그런데 roomId를 어떻게 처리해줘야 좋을지를 모르겠네요ㅜㅜ

ChoMk replied:
주석 처리한것처럼 초기화 되는 시점에 주입해서 갖고 있으면 될거 같은데

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomActivity.kt:
```diff
@@ -25,7 +25,7 @@ class ChattingRoomActivity : BaseActivity() {
             ChattingRoomViewModelInjector.provideSampleViewModelFactory()
                 .create(ChattingRoomViewModel::class.java)
```
여기서 rooId를 viewModel에 주입하는 건 어떨까? 싶은데?... 그리고 rooId가 필요한 객체(리스너나 유즈케이스)들을 주입 시점에 초기화를 해주는? 방식?
roomId는 항상 chatroomlist 페이지에서 불러오기 때문에 그런 방법이 좋지 않을까 싶어.,.

bsgreentea committed:
Implement #9 roodId 추가

bsgreentea committed:
Merge branch 'develop' of https://github.com/Basicbug/Messenger_Android into feature/#9_chatting_room

# Conflicts:
#	app/src/main/java/com/example/messenger/ui/friends/FriendListFragment.kt
#	app/src/main/java/com/example/messenger/ui/friends/FriendListViewModel.kt

bsgreentea committed:
Merge branch 'feature/#9_chatting_room_socket_impl' of https://github.com/Basicbug/Messenger_Android into feature/#9_chatting_room

# Conflicts:
#	app/src/main/java/com/example/messenger/database/message/MessageDao.kt
#	app/src/main/java/com/example/messenger/database/message/MessageDatabase.kt
#	app/src/main/java/com/example/messenger/repository/message/MessageRepository.kt
#	app/src/main/java/com/example/messenger/repository/message/MessageRepositoryImpl.kt
#	app/src/main/java/com/example/messenger/repository/model/Message.kt
#	app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomActivity.kt
#	app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomViewModel.kt
#	app/src/main/java/com/example/messenger/usecase/LoadMessagesUseCase.kt
#	app/src/main/res/layout/activity_chatting_room.xml

bsgreentea committed:
Fix #9 roomID -> roomId로 변경

bsgreentea committed:
Fix #9 빌드 에러 수정
- 상수 이름 변경
- 빌드 오류 수정

bsgreentea committed:
Implement #9 로컬 메세지 리스트 스크롤 리스너 구성
- 최소 메세지 개수(50개)를 정하고, 최소 메세지 개수 이상의 메세지가 존재할 경우 15 index에 위치하면 로컬 메세지를 더 불러오도록 구성
- delay 추가
- 상수 추가

bsgreentea commented on app/src/main/java/com/example/messenger/usecase/LoadMessagesUseCase.kt:
```diff
@@ -19,6 +21,7 @@ class LoadMessagesUseCase(
         disposables.add(
             messageRepository.getLatestFiftyMessages(roomId, from)
+                .delay(AppInfoConstants.NO_DUPLICATION, TimeUnit.MILLISECONDS)
                 .doOnSuccess {
```
@ChoMk  interval 적용을 어떻게 해야할지 감이 좀 안잡혀서 사용은 못해봤네요.. 대신 delay함수가 있길래 간단해보여서 저렇게 썼는데 한번 확인 부탁드립니다 ㅎㅎ;;

bsgreentea committed:
Test #9 채팅방 메세지 리스트 테스트
- 뭐가 문제인지 잘 모르겠네요..

bsgreentea commented on app/src/main/java/com/example/messenger/ui/chattingroom/adapter/MessageAdapter.kt:
```diff
@@ -46,8 +54,10 @@ class MessageAdapter : ListAdapter<Message, RecyclerView.ViewHolder>(
         } else if (holder is MessageSentViewHolder) {
+//            holder.bind(getItem(position))
             holder.bind(messageList[position])
         }
```
getItem(position)으로 하면 에러가 나는데, 인덱스로 사용하는것과 차이점이 있나요???

bsgreentea commented:
@ChoMk @qwebnm7788 
더 시간 끌면 안될거같아서 도움 청하려고 일단 올립니다
메세지가 리사이클러뷰에 안나타는 이유가 대체 뭘까요ㅜㅜ

bsgreentea committed:
Implement #9 테스트용 임시 코드 작성

bsgreentea committed:
Fix #9 리스트 어댑터 수정
- getItem 사용해서 정보를 가져옴

bsgreentea committed:
Implement #9 로컬 메세지 로드 구현
- ChattingRoomEvent에 메세지 리스트를 위한 PublishSubject 추가
- 로컬 메세지 로드시 리사이클러뷰의 상단에 데이터를 추가하도록 변경
- 로컬 데이터베이스에서 시간의 역순으로 메세지를 불러오기 때문에, 다시 역순으로 만들어줌

bsgreentea commented on app/src/main/java/com/example/messenger/usecase/LoadMessagesUseCase.kt:
```diff
@@ -20,11 +22,8 @@ class LoadMessagesUseCase(
         disposables.add(
             messageRepository.getLatestFiftyMessages(roomId, from)
-//                .delay(AppInfoConstants.NO_DUPLICATION, TimeUnit.MILLISECONDS)
                 .doOnSuccess {
```
리사이클러뷰 스크롤 리스너에서 요청이 발생하기 때문에 이 부분의 delay로는 해결이 불가능해서 방법을 변경했습니다.

bsgreentea commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomViewModel.kt:
```diff
@@ -57,12 +59,22 @@ class ChattingRoomViewModel(
         )
-        receiveMessageUseCase.subscribeChattingRoom("a5f4974e-bdbe-4f58-8d66-c7fd1ea4449e")
+
+        disposables.add(
+            ChattingRoomEvent.addMessagesToListSubject
+                .debounce(AppInfoConstants.NO_DUPLICATION, TimeUnit.MILLISECONDS)
+                .subscribe {
+                    messageList.value?.addAll(0, it)
+                    messageList.postValue(messageList.value)
+                }
+        )
```
ChattingRoomEvent에 리스트를 위한 PublishSubject를 추가했습니다.
debounce로 중복 로드를 방지했습니다.
잘 되네요 ㅎㅎ

ChoMk replied:
NO_DUPLICATION 이 친구는 이쪽 class로 private const val로 옮기는게 좋을거 같아... debounce가 매번 같은 시간으로 작동하지는 않을거고... 그리고 변수명은 no duplication debounce time으로 하면 좋지 않을까? 저렇게만 본다면 노 중복으로만 보면 인자가 뭐가 들어가는지 잘 모르게 그러면 디바운스를 한번 타고 들어가서 다시 확인해야할거 같아

bsgreentea commented:
@ChoMk @qwebnm7788 
첫 로드시에 리사이클러뷰의 스크롤 포지션을 최하단으로 만들려고 합니다.
ChattingRoomViewModel에 recyclerview를 주입해서 초기 로드시에 스크롤 포지션을 최하단으로 만들어주려고 했는데, 굳이 분리해둔걸 주입하는게 좀 이상한거같더라고요. 이렇게 하는게 맞는건지 다른 좋은 방법이 있을지 궁금합니다.

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/ChattingRoomActivity.kt:
```diff
@@ -0,0 +1,75 @@
+        })
+    }
+
+    override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
```
여백을 누를 때 키보드를 숨기는 친구인가? 채팅이 아니어도 다른 엑티비티에서 사용하는 경우가 많을꺼 같은데 베이스 엑티비티로 분리하는 건 어떨까 싶음

ChoMk commented on app/src/main/java/com/example/messenger/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,10 @@
+ * @author bsgreentea
+ */
+interface BaseViewModel {
+    fun setOnListener(view: View) = Unit
```
BaseViewModel의 추상화라는게 결국 viewmodel의 베이스 인데 viewmodel이 뷰에게 리스너를 달아주는게 좀 이상한거 같으 그리고 이미 basehelper라는게 있고

ChoMk replied:
helper 패턴의 의미를 한번 검색해 보면은 helper라는 친구는 사실 있어도 그만 없어도 그만인 진짜 도우미 친구인거지 즉 원래 리스너는 엑티비티에서 다는 경우가 많았으니 이를 도와주는 격으로 헬퍼에서 빼준거.. 사실 다시 지금 생각해 보니 base helper라는 의미 보다는 base activity helper가 더 좋은거 같기도 ㅎㅎ

ChoMk replied:
그리고 만약에 이 부분을 유지한다면 리사이클러뷰의 리스너만 뷰모델에서 달겠어 이런식이어야 하는데... 그게 단순히 변수명으로만은 의미전달이 힘든거 같아 그래서 나중가면은 헬퍼에서도 리스너를 달고 뷰모델에서도 리스너를 달고 그러다가 정작 리스너를 다는 공간이 분산되고... 꼬이고...ㅓ재ㅑ렂댜ㅐㅓ래쟠ㅋㅋㅋ

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/adapter/MessageRecyclerViewListener.kt:
```diff
@@ -0,0 +1,36 @@
+        val itemTotalCount = recyclerView.adapter?.itemCount
+
+        if (itemTotalCount != null) {
+
```
itemTotalCount?.let { 
            
        }

ChoMk replied:
사용해서 널체크하장

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/adapter/MessageRecyclerViewListener.kt:
```diff
@@ -0,0 +1,36 @@
+
+                if (firstVisibleItemPosition == AppInfoConstants.LOAD_MESSAGE_POINT) {
+                    loadMessagesUseCase.loadMessages(roomId, itemTotalCount)
+                }
```
MIN_SIZE_OF_MESSAGES LOAD_MESSAGE_POINT 둘다 이쪽으로 옮겨도...

ChoMk commented on app/src/main/java/com/example/messenger/ui/chattingroom/adapter/MessageRecyclerViewListener.kt:
```diff
@@ -0,0 +1,36 @@
+
+            if (itemTotalCount >= AppInfoConstants.MIN_SIZE_OF_MESSAGES) {
+
+                if (firstVisibleItemPosition == AppInfoConstants.LOAD_MESSAGE_POINT) {
```
최소 크기보다 큰 경우에 LOAD_MESSAGE_POINT위치에 도달하면 메시지를 로드 하겠다는 의미 같은데 맞나??? 한번 디버깅을 돌려봐야할거 같으데.

ChoMk replied:
돌려봤더니 시나리오는 좋음 그 리스너 이야기한 코멘트처럼 헬퍼로 분리해서 거기서 달아주면 좋을 듯 한데
viewmodel을 직접 넘기기 보다는 엑티비티를 통해서 유즈케이스만 넘기던지 아니면 엑티비티에서 유즈케이스 실행함수를 만들어서 이를 헬퍼에서 작동시킨다던지 그런 방향?

ChoMk commented:
이거 확인하면 닫고 코드리뷰 반영해서 한번 더 올려줘 그걸로 이게 길어지다 보니 읽기도 힘들고 놓지는 부분도 생기네

ChoMk commented:
다음 진행은 소켓 만들어 놓은걸로 연결만 잘 하면 될거 같은데 ㅎㅎ

bsgreentea commented:
> 이거 확인하면 닫고 코드리뷰 반영해서 한번 더 올려줘 그걸로 이게 길어지다 보니 읽기도 힘들고 놓지는 부분도 생기네

오늘 어디좀 갔다올 일이 있어서 집 오는대로 피드백 반영하고 바꿔놓겠습니당 👍
